File: frontend/types/construction.ts
export interface ConstructionProject {
  id: string
  project_name: string
  project_code: string
  location: string
  client: string
  contract_value: number
  start_date: string
  expected_completion: string
  current_progress_percentage: number
  status: 'planning' | 'active' | 'suspended' | 'completed' | 'on_hold'
  project_manager: string
  created_at?: string
}

export interface BillOfQuantities {
  id: string
  project_id: string
  boq_number: string
  item_description: string
  unit: string
  quantity: number
  unit_rate: number
  total_amount: number
  category: 'civil' | 'structural' | 'electrical' | 'plumbing' | 'finishing' | 'other'
  status: 'planned' | 'in_progress' | 'completed' | 'on_hold'
  created_at?: string
}

export interface ProgressTracking {
  id: string
  project_id: string
  date: string
  activity_description: string
  quantity_completed: number
  unit: string
  percentage_complete: number
  workforce_deployed: number
  notes: string
  photo_url?: string
  created_at?: string
}

export interface QualityControl {
  id: string
  project_id: string
  boq_item_id: string
  inspection_date: string
  inspector_name: string
  quality_status: 'passed' | 'failed' | 'partial' | 'pending'
  observations: string
  corrective_actions?: string
  follow_up_date?: string
  created_at?: string
}

export interface ConstructionEquipment {
  id: string
  project_id: string
  equipment_type: string
  equipment_name: string
  quantity: number
  deployment_date: string
  removal_date?: string
  operator_name: string
  status: 'active' | 'idle' | 'maintenance' | 'removed'
  utilization_hours: number
}

export interface ConstructionDashboard {
  total_projects: number
  active_projects: number
  avg_progress_percentage: number
  completed_projects: number
  boq_items_total: number
  boq_items_completed: number
  quality_pass_rate: number
  equipment_deployed: number
  workforce_deployed: number
  project_timeline_status: 'on_track' | 'at_risk' | 'delayed'
}

File: internal/models/construction.go
package models

import "time"

// ConstructionProject represents a construction project
type ConstructionProject struct {
	ID                     int64  `gorm:"primaryKey"`
	TenantID               string `gorm:"index"`
	ProjectName            string
	ProjectCode            string `gorm:"uniqueIndex"`
	Location               string
	Client                 string
	ContractValue          float64
	StartDate              time.Time
	ExpectedCompletion     time.Time
	CurrentProgressPercent int
	Status                 string // planning, active, suspended, completed, on_hold
	ProjectManager         string
	CreatedAt              time.Time
	UpdatedAt              time.Time
}

// BillOfQuantities represents items in the bill of quantities
type BillOfQuantities struct {
	ID              int64  `gorm:"primaryKey"`
	TenantID        string `gorm:"index"`
	ProjectID       int64
	BOQNumber       string
	ItemDescription string
	Unit            string
	Quantity        float64
	UnitRate        float64
	TotalAmount     float64
	Category        string // civil, structural, electrical, plumbing, finishing, other
	Status          string // planned, in_progress, completed, on_hold
	CreatedAt       time.Time
	UpdatedAt       time.Time
}

// ProgressTracking represents project progress records
type ProgressTracking struct {
	ID                int64  `gorm:"primaryKey"`
	TenantID          string `gorm:"index"`
	ProjectID         int64
	Date              time.Time
	ActivityDesc      string `gorm:"type:text"`
	QuantityCompleted float64
	Unit              string
	PercentComplete   int
	WorkforceDeployed int
	Notes             string `gorm:"type:text"`
	PhotoURL          string
	CreatedAt         time.Time
	UpdatedAt         time.Time
}

// QualityControl represents quality control inspections
type QualityControl struct {
	ID               int64  `gorm:"primaryKey"`
	TenantID         string `gorm:"index"`
	ProjectID        int64
	BOQItemID        int64
	InspectionDate   time.Time
	InspectorName    string
	QualityStatus    string // passed, failed, partial, pending
	Observations     string `gorm:"type:text"`
	CorrectiveAction string `gorm:"type:text"`
	FollowUpDate     time.Time
	CreatedAt        time.Time
	UpdatedAt        time.Time
}

// ConstructionEquipment represents equipment used in construction
type ConstructionEquipment struct {
	ID             int64  `gorm:"primaryKey"`
	TenantID       string `gorm:"index"`
	ProjectID      int64
	EquipmentName  string
	EquipmentType  string
	SerialNumber   string
	Status         string // available, in_use, maintenance, retired
	DeploymentDate time.Time
	RetirementDate time.Time
	CostPerDay     float64
	Notes          string `gorm:"type:text"`
	CreatedAt      time.Time
	UpdatedAt      time.Time
}

// TableName specifies the table name for ConstructionProject
func (ConstructionProject) TableName() string {
	return "construction_projects"
}

// TableName specifies the table name for BillOfQuantities
func (BillOfQuantities) TableName() string {
	return "bill_of_quantities"
}

// TableName specifies the table name for ProgressTracking
func (ProgressTracking) TableName() string {
	return "progress_tracking"
}

// TableName specifies the table name for QualityControl
func (QualityControl) TableName() string {
	return "quality_control"
}

// TableName specifies the table name for ConstructionEquipment
func (ConstructionEquipment) TableName() string {
	return "construction_equipment"
}
